<a href="https://colab.research.google.com/github/johacagua-mina/JohannaNova/blob/main/Task3_MSO4130.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MSO4130 Portfolio Task 3
**Student:** Johanna Cagua Mina  
**Student ID:** M01089122

## Objective
This notebook answers Portfolio Task 3 using Python. The task uses logistic regression to build a credit rating model for predicting default risk.

In [1]:
import pandas as pd
import statsmodels.api as sm
from google.colab import files

In [2]:
uploaded = files.upload()

Saving credit_data_student_M01089122 (1).csv to credit_data_student_M01089122 (1).csv


In [7]:
df = pd.read_csv("credit_data_student_M01089122 (1).csv")
df.columns

Index(['Annual_Income', 'Debt_to_Income', 'Credit_Utilization',
       'Recent_Late_Payments', 'Years_Employed', 'Default'],
      dtype='object')

In [9]:
df = pd.read_csv("credit_data_student_M01089122 (1).csv")
df.head()

,Annual_Income,Debt_to_Income,Credit_Utilization,Recent_Late_Payments,Years_Employed,Default
0,67371.17,0.230,0.110,2,1.2,0
1,62945.56,0.417,0.041,0,2.6,1
2,71927.91,0.432,0.307,1,0.4,1
3,65266.53,0.128,0.556,0,5.1,0
4,47899.28,0.330,0.265,1,2.2,0


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Annual_Income         1000 non-null   float64
 1   Debt_to_Income        1000 non-null   float64
 2   Credit_Utilization    1000 non-null   float64
 3   Recent_Late_Payments  1000 non-null   int64  
 4   Years_Employed        1000 non-null   float64
 5   Default               1000 non-null   int64  
dtypes: float64(4), int64(2)
memory usage: 47.0 KB


In [10]:
X = df[["Annual_Income", "Debt_to_Income", "Credit_Utilization", "Recent_Late_Payments", "Years_Employed"]]
y = df["Default"]

X = sm.add_constant(X)

logit_model = sm.Logit(y, X).fit()
print(logit_model.summary())

Optimization terminated successfully.
         Current function value: 0.658966
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:                Default   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      994
Method:                           MLE   Df Model:                            5
Date:                Tue, 17 Mar 2026   Pseudo R-squ.:                 0.03964
Time:                        15:33:55   Log-Likelihood:                -658.97
converged:                       True   LL-Null:                       -686.17
Covariance Type:            nonrobust   LLR p-value:                 1.730e-10
                           coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                   -0.3093      0.332     -0.930      0.352      -0.961       0.342

## Part (b) Logistic Regression and Credit Rating

A logistic regression model was used to estimate the probability that a customer defaults on their loan.

The estimated credit rating (log-odds of default) is given by:

Credit Rating = -0.3093
- 1.387e-05 × Annual_Income
+ 1.4324 × Debt_to_Income
+ 1.3058 × Credit_Utilization
+ 0.3597 × Recent_Late_Payments
- 0.0460 × Years_Employed

The coefficients indicate how each variable affects the likelihood of default.

Higher income and longer employment reduce the probability of default, while higher debt-to-income ratio, credit utilisation, and recent late payments increase the probability of default.

In particular, recent late payments have a strong positive effect, suggesting that past repayment behaviour is one of the most important predictors of default risk.

P(default) > 50%

Credit Rating = 0

A probability threshold of 50% corresponds to a credit rating of 0.

Therefore, customers with a credit rating greater than 0 should be rejected, as they have a predicted probability of default greater than 50%.

## Part (c) Interpretation of Annual Income

Although the coefficient for Annual Income appears relatively small, this does not mean that income is unimportant.

One reason is that other variables in the model, such as Debt_to_Income and Credit_Utilization, already capture aspects of a customer's financial situation that are closely related to income.

In particular, the debt-to-income ratio directly incorporates income, meaning that the effect of income is partly absorbed by this variable.

Furthermore, variables such as Recent_Late_Payments reflect actual repayment behaviour, which is often a stronger predictor of default than income alone.

Therefore, while income is relevant, its independent effect appears smaller once we control for more behaviour-based measures of financial risk.

## Part (d) New Rejection Threshold

For a logistic regression model, the probability of default is linked to the credit rating through the logit transformation:

z = ln(p / (1-p))

If the bank wants to reject applications when the probability of default is greater than 20%, we set p = 0.20:

z = ln(0.20 / 0.80)
z = ln(0.25)
z ≈ -1.386

Therefore, the new rejection threshold for the credit rating is approximately -1.386.

This means that loan applications should be rejected whenever the credit rating is greater than -1.386.

This lower threshold reflects a more risk-averse lending strategy, as the bank is willing to reject customers at a lower predicted probability of default. As a result, more applicants will be rejected compared to the 50% threshold.